deep_qa/data/dataset_readers/language_modeling.py

```diff
@@ -0,0 +1,111 @@
+        one with default parameters.
+    """
+    def __init__(self,
+                 filename: str,
+                 tokens_per_instance: int=None,
+                 tokenizer: Tokenizer=WordTokenizer(),
+                 token_indexers: List[TokenIndexer]=None):
+        self._filename = filename
+        self._tokens_per_instance = tokens_per_instance
+        self._tokenizer = tokenizer
+        if token_indexers is None:
+            token_indexers = [SingleIdTokenIndexer()]
+        self._token_indexers = token_indexers
+        self._start_token = '<S>'
+        self._end_token = '</S>'
+
+    @overrides
+    def read(self):
+        with open(self._filename, "r") as text_file:
+            instance_strings = text_file.readlines()
+        if self._tokens_per_instance is not None:
+            all_text = " ".join([x.replace("\n", " ").strip() for x in instance_strings])
+            tokenized_text = self._tokenizer.tokenize(all_text)
+            num_tokens = self._tokens_per_instance
+            tokenized_strings = []
+            for index in range(0, len(tokenized_text) - num_tokens, num_tokens):
+                tokenized_strings.append(tokenized_text[index:index + num_tokens])
+        else:
+            tokenized_strings = [self._tokenizer.tokenize(s) for s in instance_strings]
+
+        # TODO(matt): this isn't quite right, because you really want to split on sentences,
+        # tokenize the sentences, add the start and end tokens per sentence, then change the tokens
+        # per instance if desired.  But, we can fix that later, if someone actually wants to use
+        # this for language modeling.  This is just another example of how to use the data reader
+        # code, for now.
+        tokenized_strings = [[self._start_token] + x + [self._end_token] for x in tokenized_strings]
+
+        # No matter how you want to represent the input, we'll always represent the output as a
+        # single token id.  This code lets you learn a language model that concatenates word
+        # embeddings with character-level encoders, in order to predict the word token that comes
+        # next.
+        output_indexer = None
+        for indexer in self._token_indexers:
+            if isinstance(indexer, SingleIdTokenIndexer):
+                output_indexer = indexer
+                break
+        else:
+            output_indexer = SingleIdTokenIndexer()
+
+        instances = []
+        for tokenized_string in tokenized_string:
+            input_field = TokenizedTextField(tokenized_string[:-1], self._token_indexers)
+            output_field = TokenizedTextField(tokenized_string[1:], [output_indexer])
+            instances.append(Instance({'input_tokens': input_field,
+                                       'output_tokens': output_field}))
+        return Dataset(instances)
+
+    @classmethod
+    def from_params(cls, params: Params):
```

Comment:
There will be a `from_params` method on the abstract base class, like for [`TokenIndexer`](https://github.com/allenai/deep_qa/pull/402/files#diff-ea26f15c9581a597a609b713eaf11965R79) and [`Tokenizer`](https://github.com/allenai/deep_qa/pull/402/files#diff-ca9750b4d8885ba66736abff077b9722R32), that does what you put in your code block - I just forgot to implement it for `DatasetReader`.  So the code that constructs a `DatasetReader` from parameters actually does just look like the block I showed:

```python
dataset_reader_params = params.pop('dataset')
dataset_reader = DatasetReader.from_params(dataset_reader_params)
```

The difference between separating out the use of `Params` from the constructor is two-fold:

1. It makes it more obvious what the actual objects are that you're taking as a dependency.  When you're looking at the constructor, you can see the right object types for, e.g., the tokenizer, instead of just getting a string description of the tokenizer.  This also helps someone new to the codebase who is familiar with python get started easier.

2. It separates the experiment framework code (`Params`) from the data generator code.  If you want to totally ignore our experiment framework, but still use our data processing and other code, this makes it a lot easier.